src/scale.jl

```diff
@@ -251,45 +253,51 @@ function apply_scale_typed!(ds, field, scale::ContinuousScale)
     permute!(level_values, order)
-    return PooledDataArray(da, level_values)
+    return IndirectArray(da, level_values)
 end
 
-function discretize_make_pda(values::Vector, levels=nothing)
-    if levels == nothing
-        return PooledDataArray(values)
-    else
-        return PooledDataArray(convert(Vector{eltype(levels)}, values), levels)
-    end
-end
+discretize_make_pda(values::Vector)         = IndirectArray(values)
+discretize_make_pda(values::Vector, ::Void) = IndirectArray(values)
+discretize_make_pda(values::Vector, levels) =
+    IndirectArray(map(t -> findfirst(levels, t), values), levels)
 
-function discretize_make_pda(values::DataArray, levels=nothing)
-    if levels == nothing
-        return PooledDataArray(values)
-    else
-        return PooledDataArray(convert(DataArray{eltype(levels)}, values), levels)
-    end
-end
+discretize_make_pda(values::DataArray)         = IndirectArray(values)
+discretize_make_pda(values::DataArray, ::Void) = IndirectArray(values)
+discretize_make_pda(values::DataArray, levels) =
+    IndirectArray(convert(DataArray{eltype(levels)}, values), levels)
 
-function discretize_make_pda(values::Range, levels=nothing)
-    if levels == nothing
-        return PooledDataArray(collect(values))
-    else
-        return PooledDataArray(collect(values), levels)
-    end
-end
+discretize_make_pda(values::Range)         = IndirectArray(collect(values))
+discretize_make_pda(values::Range, ::Void) = IndirectArray(collect(values))
+discretize_make_pda(values::Range, levels) = IndirectArray(collect(values), levels)
 
-function discretize_make_pda(values::PooledDataArray, levels=nothing)
-    if levels == nothing
-        return values
-    else
-        return PooledDataArray(values, convert(Vector{eltype(values)}, levels))
-    end
+discretize_make_pda(values::IndirectArray)         = values
+discretize_make_pda(values::IndirectArray, ::Void) = values
+discretize_make_pda(values::IndirectArray, levels) =
+    IndirectArray(values, convert(Vector{eltype(values)}, levels))
+
+discretize_make_pda(values::CategoricalArray)         = discretize_make_pda(values, unique(values))
+discretize_make_pda(values::CategoricalArray, ::Void) = discretize_make_pda(values)
+function discretize_make_pda(values::CategoricalArray{T}, levels::Vector) where {T}
+    index = map!(t -> ismissing(t) ?
+                    findfirst(ismissing, levels) :
+                        findfirst(s -> !(ismissing(s) || s != get(t)), levels),
+                 Array{UInt8}(size(values)...), values)
+    return IndirectArray(index, convert(Vector{T},levels))
 end
+function discretize_make_pda(values::CategoricalArray{T}, levels::CategoricalVector{T}) where T
+    _levels = map!(t -> ismissing(t) ? t : get(t), Vector{T}(length(levels)), levels)
+    discretize_make_pda(values, _levels)
+end
+
+# These methods convert WeakRefStringArrays to Vector{String} and shouldn't really be necessary
```

Comment:
Should be OK with recent releases of WeakRefStrings, right?

Reply:
Probably not, but it could happen so I think we should keep them.

Reply:
`WeakRefStringArray`s now return `String` objects when indexed for several releases, that's why I said that's probably no longer needed (as the comment says).

Reply:
`WeakRefStringArrays` are still passed to Gadfly and no other signature currently matches.

Reply:
How about changing the `::Vector` method to `::AbstractVector`, and using the `IndirectArray` convenience constructor you've just added? That would allow getting rid of the dependencies on WeakRefStrings and DataArrays, which really shouldn't be here since we only use the `AbstractArray` interface.

Reply:
Again, we don't want any wrapped types to escape to `Compose` which can be difficult when using `AbstractArray` signatures. Also, note that there isn't a dependency on `WeakRefStrings `. I'm fine with working towards dropping the `DataArrays` dependency but not in this PR.